### DISCUSSION TITLE
feature: Improve documentation (for formatting) for newbies

### DISCUSSION THREAD

**Discussion by tf42src** [LazyVim/LazyVim#3144] [Q&A]
### Did you check the docs?

- [X] I have read all the LazyVim docs

### Is your feature request related to a problem? Please describe.

I am trying to format html. It is not fully clear to me what the required steps are. Do I need to install conform.nvim? Do I need to install a language server, do I need to have npm, do I need prettier?

I started with the [StarterTemplate](https://github.com/LazyVim/starter) which works great. I am really enjoying it and I am getting used more and more how to use everything. What I couldn't figure out so far what steps do I need to do to enable formatting (on save and on shortcut) for html. 

`<leader>cf` (while a html file is open) currently gives me "No formatter available". So I head over to [Formatting](https://www.lazyvim.org/plugins/formatting) which tells me I should use `conform.nvim` - which seems to be installed (see screenshot).

[Formatting](https://www.lazyvim.org/plugins/formatting), also gives some configuration options: `opts` and I kind of understand that I need to add sth. like `html = {"prettier"}` to `formatters_by_ft`. Makes sense, but where? 

I cannot find `conform.nvim` anywhere in my `.config/nvim/` folder except for one entry in `./lazy-lock.json`. I followed this [well-explained youtube video](https://www.youtube.com/watch?v=ybUE4D80XSk&t=772s&pp=ygUSbGF6eXZpbSBmb3JtYXR0aW5n) to enable formatting which lead to "Don't override plugin.config directly, since this will break LazyVim formatting."

I also tried copying the "Full Spec" from [Formatting](https://www.lazyvim.org/plugins/formatting) into a new file in my plugins directory which also didn't work (the built-in lua LSP hinted at all sorts of errors).

Apologies for the lengthy description, but hopefully it makes it a bit clear what a new user of lazyvim might go through and struggle with.

![Screenshot from 2024-05-11 09-59-38](https://github.com/LazyVim/LazyVim/assets/9712861/5456bb0d-df25-49a8-8bc9-f6de6f317b4d)

### Describe the solution you'd like

A simple example how/where to enable formatting for anything but what is enabled already.

### Describe alternatives you've considered

Youtube :)

### Additional context
```
NVIM v0.9.4
Build type: RelWithDebInfo
LuaJIT 2.1.1692716794
Compilation: /usr/bin/cc -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wvla -Wdouble-promotion -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-prototypes -fno-common -Wno-unused-result -Wimplicit-fallthrough -fdiagnostics-color=always -fstack-protector-strong -DUNIT_TESTING -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/build/nvim/parts/nvim/build/.deps/usr/include/luajit-2.1 -I/usr/include -I/build/nvim/parts/nvim/build/.deps/usr/include -I/build/nvim/parts/nvim/build/build/src/nvim/auto -I/build/nvim/parts/nvim/build/build/include -I/build/nvim/parts/nvim/build/build/cmake.config -I/build/nvim/parts/nvim/build/src -I/usr/include -I/build/nvim/parts/nvim/build/.deps/usr/include -I/build/nvim/parts/nvim/build/.deps/usr/include -I/build/nvim/parts/nvim/build/.deps/usr/include -I/build/nvim/parts/nvim/build/.deps/usr/include -I/build/nvim/parts/nvim/build/.deps/usr/include -I/build/nvim/parts/nvim/build/.deps/usr/include

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info

```

**Comment by dpetka2001** [comment]
> [Formatting](https://www.lazyvim.org/plugins/formatting), also gives some configuration options: opts and I kind of understand that I need to add sth. like html = {"prettier"} to formatters_by_ft. Makes sense, but where?
I cannot find conform.nvim anywhere in my .config/nvim/ folder except for one entry in ./lazy-lock.json. I followed this [well-explained youtube video](https://www.youtube.com/watch?v=ybUE4D80XSk&t=772s&pp=ygUSbGF6eXZpbSBmb3JtYXR0aW5n) to enable formatting which lead to "Don't override plugin.config directly, since this will break LazyVim formatting."

It's mentioned in the docs where you should place your own customization right [here](https://www.lazyvim.org/configuration/plugins#-adding-plugins). So you just create a file in `/lua/plugins/` directory and inside the file you make the changes you want for the specific plugin spec you're interested in. So, you would do something like
```lua
return {
{
  "stevearc/conform.nvim",
  opts = {
    formatters_by_ft = {
      ["html"] = { "prettier" },
    },
  },
},
}
```
> I also tried copying the "Full Spec" from [Formatting](https://www.lazyvim.org/plugins/formatting) into a new file in my plugins directory which also didn't work (the built-in lua LSP hinted at all sorts of errors).

You should never copy the entire from what's in the docs. Usually there's a warning about this in the docs, but for some reason in the formatting page it's missing.
> caution
You don't need to copy the default settings to your config. They are only shown here for reference.

So, in summary, you create files in `~/.config/nvim/lua/plugins/` directory and there you make changes to the specs of the plugins you want to change, as is mentioned in the [docs](https://www.lazyvim.org/configuration/plugins#-adding-plugins).

**Comment by tf42src** [comment]
Thanks for the quick reply! I have added the "opts" in my plugin directory like this:

![Screenshot from 2024-05-12 01-22-36](https://github.com/LazyVim/LazyVim/assets/9712861/95ca24db-f9ce-452b-a370-66ac5265f442)

(I have removed the "format_on_save" again - because that wasn't allowed). Nevertheless, navigating to an html file and pressing `<leader>cf` still shows, I don't have a formatter:

![Screenshot from 2024-05-12 01-22-20](https://github.com/LazyVim/LazyVim/assets/9712861/b71606f0-fd9e-4647-a201-e528cf03ed6f)

I have double checked, that `prettier` is installed.

![Screenshot from 2024-05-12 01-28-17](https://github.com/LazyVim/LazyVim/assets/9712861/59d80f17-b13b-4429-8454-3456b376085c)

The only hint I could find was this (but it doesn't tell me what the error is):

```
cat  ~/.local/state/nvim/conform.log
01:23:21[ERROR] Formatter 'prettier' error: 
01:23:38[ERROR] Formatter 'prettier' error: 
01:23:40[ERROR] Formatter 'prettier' error: 
01:23:45[ERROR] Formatter 'prettier' error: 
```

I can run prettier manually and it works:

`~/node_modules/.bin/prettier ~/projects/testproj/intro.html`

**Comment by tf42src** [comment]
Did a bit more digging and found [this](https://github.com/LazyVim/LazyVim/discussions/3096), which made me check what I have installed, but everything looks fine:

![Screenshot from 2024-05-12 01-54-09](https://github.com/LazyVim/LazyVim/assets/9712861/cef0e264-db2c-490d-b7e6-1de64a7cd121)

**Comment by dpetka2001** [comment]
What does `:!which prettier` show inside Neovim?

**Comment by tf42src** [comment]
![Screenshot from 2024-05-12 15-53-02](https://github.com/LazyVim/LazyVim/assets/9712861/1581fe34-f1c2-4439-a268-43bcb1143c0f)

I can execute `prettier` directly from the command line too (it's added to `$PATH`).

**Comment by dpetka2001** [comment]
What does `LazyFormatInfo` and `ConformInfo` show?

**Comment by tf42src** [comment]
Looks kinda ok, except for the errors I posted above already. Unfortunately it doesn't tell me what error "prettier" has. Also if I clear that file, then try to `<leader>cf` again, there are no errors in `.local/state/nvim/conform.log`.

Can I see the exact command nvim is executing somewhere? To reproduce it outside of the editor.

![Screenshot from 2024-05-13 11-10-28](https://github.com/LazyVim/LazyVim/assets/9712861/822e2bfc-2e66-4ab5-b8d4-2fd87b7cda14)
![Screenshot from 2024-05-13 12-08-01](https://github.com/LazyVim/LazyVim/assets/9712861/f309ef95-7cac-4f73-a553-724d3ea2212f)

**Reply by tf42src** [reply]
I tried replacing "prettier" with "djlint" which works for sth. simple like this:

**plain.html**
```
<html>
    <head></head>
    <body>foobar</body>
</html>
```

It does not work for larger files though. Now it is timing out, and again, it doesn't allow me to update `format_on_save`. :(

![Untitled](https://github.com/LazyVim/LazyVim/assets/9712861/88579194-c7ce-4c30-a197-3c3e4cd7c76e)

At least for the small sample file (plain.html) I do not get the "No formatters available for this buffer" anymore, but for another I do!
I will investigate more and post another update.

**Reply by tf42src** [reply]
Yeah, very puzzled by the issue. I have two files which seem to be identical, but one works - and the other one doesn't.

**Video showing the issue**
[Screencast from 13-05-24 14:08:11.webm](https://github.com/LazyVim/LazyVim/assets/9712861/29e40a12-250b-43d0-b9c3-491d2f48aaf9)

![Screenshot from 2024-05-13 14-18-28](https://github.com/LazyVim/LazyVim/assets/9712861/4cca4ad4-7f6d-41f3-95c3-aa101a016748)

**Reply by tf42src** [reply]
If I close and reopen the buffer, it works. How does lazyvim/nvim select a formatter? Solely by file type or does it check _anything_ else? What could lead to the "No formatters available for this buffer" message?

**Reply by dpetka2001** [reply]
What does `:set ft?` show in the buffer you're having problems with? Because in your previous screenshot it shows that `ConformInfo` doesn't find a formatter for the buffer you're in, but `prettier` is configured for `html`.

**Reply by tf42src** [reply]
![Screenshot from 2024-05-14 10-14-28](https://github.com/LazyVim/LazyVim/assets/9712861/922e5f53-f121-427c-9a0e-9da3ac497226)

That's it! It was a django (jinja2) file before. I had cleaned it up to see if that makes any change. Seems lazy/nvim remembered what file was and does not update the classification as the file changes. I have added:

```
htmldjango = { "djhtml" }
``` 
to the opts in the config, restarted nvim and I can format the file now.

Thanks for being so responsive!

**Reply by tf42src** [reply]
I also figured out how to fix the timeout issue, had to use `format` instead of `format_on_save`. Now everything works nicely. Looking forward to my new IDE. 

Thanks again!
